### PR TITLE
Improve documentation of how JSON.stringify serializes things

### PR DESCRIPTION
**Pull request by gzm0**
This is in response to #4094.

### CODE REVIEW

**Comment by gzm0** [comment]
@sjrd it might be a good idea to hold off on this until `1.1.1` is out? (I'm working on a fix on #4085, but the testing is a bit involved :)).

File: library/src/main/scala/scala/scalajs/js/JSON.scala [hunk 1]
```diff
@@ -46,6 +46,26 @@ object JSON extends js.Object {
...
+   *     found in an [[Array]]).
+   *   - All [[Symbol]]-keyed properties will be completely ignored, even when
+   *     using the replacer function.
+   *   - The instances of [[Date]] implement the [[Date#toJSON()*]] function by
+   *     returning a string (the same as [[Date#toISOString]]). Thus, they are
+   *     treated as strings.
+   *   - The numbers Infinity and NaN, as well as the value null, are all
+   *     considered null.
+   *   - All the other [[Object]] instances (including Map, Set, WeakMap, and
+   *     WeakSet) will have only their enumerable properties serialized.
```
Comment (sjrd) [inline library/src/main/scala/scala/scalajs/js/JSON.scala:67]: We should add a paragraph that explains what happens when the top-level argument is a function or `undefined`. Basically a replacement for the MDN doc that says:

> JSON.stringify() can return undefined when passing in "pure" values like JSON.stringify(function(){}) or JSON.stringify(undefined).

In our case we should say that those calls are invalid and will result in Undefined Behavior.

File: library/src/main/scala/scala/scalajs/js/JSON.scala [hunk 2]
```diff
@@ -46,6 +46,26 @@ object JSON extends js.Object {
    * is specified, or optionally including only the specified properties if a
    * replacer array is specified.
    *
+   *   - If the value has a
+   *     [[https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#toJSON_behavior toJSON()]]
+   *     method, it's responsible to define what data will be serialized.
```
Comment (sjrd) [inline library/src/main/scala/scala/scalajs/js/JSON.scala:51]:
```suggestion
   *     method, it is responsible to define what data will be serialized.
```

**Comment by sjrd** [comment]
> @sjrd it might be a good idea to hold off on this until `1.1.1` is out? (I'm working on a fix on #4085, but the testing is a bit involved :)).

Yes, I agree that we should release a 1.1.1 before merging this PR.

**Comment by gzm0** [comment]
I have isolated the commit that just improves the doc so we can merge it now.

**Comment by gzm0** [comment]
@sjrd any further comments on this? Some of the details changed to make CI happy.

**Pull request merged**